src/Api/LockFreeApiEventProducer.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "include/Api/LockFreeApiEventProducer.h"
+
+namespace orbit_api {
+namespace {
+template <class... Ts>
+struct overloaded : Ts... {
+  using Ts::operator()...;
+};
+template <class... Ts>
+overloaded(Ts...)->overloaded<Ts...>;
```

Comment:
Wow, the effort we have to put in to dispatch based on types is remarkable. I find this and the dispatch code below not very readable. I wonder if we could make this simpler. Could we for example have a ToCaptureEventProto(...) method in each event type that would fill a passed in CaptureEvent proto?

The dispatching code would become: 

std::visit([](auto& event) {  event.ToCaptureEvent(); }, raw_api_event);

Without the need for the variadic template magic. I feel it would be more readable to have the proto conversion code encapsulated in the event type, wdyt?

Reply:
We could also call overloaded functions from the lambda if we don't want to add a dependency to grcp protos in the interface...

Reply:
I was also unsure about that. I kind of liked that is was so compact, but I went with a set of overloaded functions in the anonymous  namespace.

Reply:
Cool, thanks!